service.go

```diff
@@ -0,0 +1,36 @@
+package hcsshim
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+func GetHighestSupportedHcsSchemaVersion() (schemaVersion *hcsschema.Version, err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Minute)
+	defer cancel()
+
+	serviceProperties, err := hcs.GetServiceProperties(ctx, hcsschema.PropertyQuery{PropertyTypes: []string{"Basic"}})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve HCS schema version: %s", err)
+	}
+
+	basicInfo := &hcsschema.BasicInformation{}
+	if err := json.Unmarshal(serviceProperties.Properties[0], &basicInfo); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal HCS Schema Version: %s", err)
+	}
+
+	for j := 1; j < len(basicInfo.SupportedSchemaVersions); j++ {
+		if basicInfo.SupportedSchemaVersions[0].Major < basicInfo.SupportedSchemaVersions[j].Major {
+			basicInfo.SupportedSchemaVersions[0] = basicInfo.SupportedSchemaVersions[j]
+		}
+	}
+
+	schemaVersion = &basicInfo.SupportedSchemaVersions[0]

```

Comment:
need to consider both major and minor versions.

 var highestVersion hcsschema.Version
 for _, version := range basicInfo.SupportedSchemaVersions {
  if version.Major >= highestVersion.Major {
   if (version.Major > highestVersion.Major) || (version.Minor > highestVersion.Minor) {
    highestVersion.Major = version.Major
    highestVersion.Minor = version.Minor
   }
  }
 }
 schemaVersion = &highestVersion

Reply:
dont think minor version need to be considered: here's the comment from Schema.Responses.Service.mars
class BasicInformation
{
    //
    // The supported schema versions will be returned as an array.
    // Array element **A.X implies all versions with major version A and minor version from 0 to X are also supported**.
    //
    [NewIn("2.1"), OmitEmpty] Schema.Version SupportedSchemaVersions[];
};

so i dont see why the array would be populated with entries where only minor version is different...

Reply:
adding the check is redundant but doesnt hurt. to save time, will check for minor version as well..

Reply:
of course that would be the case where all versions are populated. nothing to argue there..., but here 2.3 implies 23, 2.2, 2.1, 2.0 is supported. so why would they populate 2.2, 2.1 etc... my observation is on win11 it returns 2.7 and rs5 it turns up with 2.1. i havent seen it reporting 2.7,2.3,2.1 for win11...